File: jongsik_kim/2023-05-14-balanced-binary-tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def get_depth(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            return max(get_depth(node.left), get_depth(node.right)) + 1
        if not root:
            return True
        return abs(get_depth(root.left) - get_depth(root.right)) <= 1

File: wonsang-yoon/2023-05-14-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

/*
    Logic: find maximum height of the left and rigt node
    check the difference between left and right height
    if the difference is over 1, return false which is not balanced tree

*/
var isBalanced = function(root) {
    if(root === null) {return true}
    let l_node, r_node;
    //Find node Heights
    l_node = dfs(root.left);
    r_node = dfs(root.right);

    //Find Left and Right node difference
    if(Math.abs(l_node - r_node) > 1)
        return false;
    else
        return true
};
const dfs = (root) =>{
    if(!root){return 0}
    //add one every connection
    return 1+ Math.max(dfs(root.right),dfs(root.left))
}